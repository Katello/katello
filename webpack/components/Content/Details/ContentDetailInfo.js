import React from 'react';
import PropTypes from 'prop-types';
import { Table } from 'react-bootstrap';

// using Map to preserve order

const createRows = (details, mapping) => {
  const rows = [];
  /* eslint-disable no-restricted-syntax, react/jsx-closing-tag-location */
  for (const key of mapping.keys()) {
    rows.push(<tr key={key}>
      <td><b>{mapping.get(key)}</b></td>
      <td>{Array.isArray(details[key]) ? details[key].join(', ') : details[key]}</td>
    </tr>);
  }
  /* eslint-enable no-restricted-syntax, react/jsx-closing-tag-location */
  return rows;
};

const ContentDetailInfo = ({ contentDetails, displayMap }) => (
  <Table>
    <tbody>
      {createRows(contentDetails, displayMap)}
    </tbody>
  </Table>
);

ContentDetailInfo.propTypes = {
  contentDetails: PropTypes.shape({}).isRequired,
  displayMap: PropTypes.instanceOf(Map).isRequired,
};

export default ContentDetailInfo;
