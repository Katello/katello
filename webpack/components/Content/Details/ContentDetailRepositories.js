import React from 'react';
import PropTypes from 'prop-types';
import { Table } from 'patternfly-react';
import TableSchema from './ContentDetailRepositoryTableSchema';

const ContentDetailRepositories = ({ repositories }) => (
  <div>
    <Table.PfProvider columns={TableSchema}>
      <Table.Header />
      <Table.Body rows={repositories} rowKey="id" />
    </Table.PfProvider>
  </div>
);

ContentDetailRepositories.propTypes = {
  repositories: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

export default ContentDetailRepositories;
