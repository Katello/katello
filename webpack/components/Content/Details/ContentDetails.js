import React from 'react';
import { translate as __ } from 'foremanReact/common/I18n';
import { Nav, NavItem, TabPane, TabContent, TabContainer, Grid, Row, Col } from 'patternfly-react';
import { PropTypes } from 'prop-types';
import { LoadingState } from '../../../move_to_pf/LoadingState/index';

const ContentDetails = (props) => {
  const { contentDetails, schema } = props;
  const { loading } = contentDetails;

  const tabHeaders = () => {
    const tabs = schema.map(node => (
      <NavItem key={node.key} eventKey={node.key}>
        <div>{node.tabHeader}</div>
      </NavItem>
    ));
    return tabs;
  };

  const tabPanes = () => {
    const tabPane = schema.map(node => (
      <TabPane key={node.key} eventKey={node.key}>
        <Row>
          <Col sm={12}>
            {node.tabContent}
          </Col>
        </Row>
      </TabPane>
    ));
    return tabPane;
  };

  return (
    <div>
      <LoadingState
        loading={loading}
        loadingText={__('Loading')}
      >
        <TabContainer id="content-tabs-container" defaultActiveKey={1}>
          <Grid bsClass="container-fluid">
            <Row>
              <Col sm={12}>
                <Nav id="content-nav-container" bsClass="nav nav-tabs">
                  {schema && tabHeaders()}
                </Nav>
              </Col>
            </Row>
            <TabContent animation={false}>
              {schema && tabPanes()}
            </TabContent>
          </Grid>
        </TabContainer>
      </LoadingState>
    </div>
  );
};

ContentDetails.propTypes = {
  contentDetails: PropTypes.shape({
    loading: PropTypes.bool,
    name: PropTypes.string,
    // Disabling rule as existing code failed due to an eslint-plugin-react update
    /* eslint-disable react/forbid-prop-types */
    profiles: PropTypes.array,
    repositories: PropTypes.array,
    artifacts: PropTypes.array,
    /* eslint-enable react/forbid-prop-types */
    stream: PropTypes.string,
  }).isRequired,
  schema: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

export default ContentDetails;
