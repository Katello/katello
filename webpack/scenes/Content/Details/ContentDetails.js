import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import { Grid, GridItem, TextContent, Text, TextVariants, Flex, Breadcrumb, BreadcrumbItem } from '@patternfly/react-core';
import { STATUS } from 'foremanReact/constants';
import ContentConfig from '../ContentConfig';
import RoutedTabs from '../../../components/RoutedTabs';
import { selectContentDetails, selectContentDetailsStatus } from '../ContentSelectors';
import { getContentDetails } from '../ContentActions';
import Loading from '../../../components/Loading';

const ContentDetails = () => {
  const dispatch = useDispatch();
  const contentDetailsResponse = useSelector(selectContentDetails);
  const contentDetailsResponseStatus = useSelector(selectContentDetailsStatus);

  const { id, content_type: contentType } = useParams();
  const contentId = Number(id);
  const config = ContentConfig.find(type =>
    type.names.pluralLabel === contentType);
  const { pluralTitle, pluralLabel } = config.names;

  const tabs = [];
  config.tabs.forEach((tab) => {
    tabs.push({
      title: tab.title,
      key: tab.tabKey,
      content: tab.getContent(contentType, contentId, tab.tabKey),
    });
  });

  useEffect(() => {
    dispatch(getContentDetails(contentType, contentId));
  }, [dispatch, contentType, contentId]);

  if (contentDetailsResponseStatus === STATUS.PENDING) {
    return <Loading />;
  }

  return (
    <Grid>
      <Grid className="margin-16-24">
        <Breadcrumb>
          <BreadcrumbItem
            aria-label="content_breadcrumb"
            render={() => (<Link to={`/content/${pluralLabel}`}>{pluralTitle}</Link>)}
          />
          <BreadcrumbItem
            aria-label="content_breadcrumb_content"
            isActive
          > {contentDetailsResponse.name}
          </BreadcrumbItem>
        </Breadcrumb>
        <GridItem span={12} className="margin-top-24">
          <Flex>
            <TextContent>
              <Text component={TextVariants.h1}> {contentDetailsResponse.name} </Text>
            </TextContent>
          </Flex>
        </GridItem>
      </Grid>
      <GridItem span={12}>
        <RoutedTabs tabs={tabs} baseUrl={`/${contentType}/${contentId}`} defaultTabIndex={0} />
      </GridItem>
    </Grid >
  );
};
export default ContentDetails;
