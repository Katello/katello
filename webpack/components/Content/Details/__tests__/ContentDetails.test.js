import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
// import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';
import ContentDetailInfo from '../ContentDetailInfo';
import ContentDetailRepositories from '../ContentDetailRepositories';
import ContentDetails from '../ContentDetails';

jest.mock('foremanReact/components/Pagination/PaginationWrapper', () => (<div>Pagination Mock</div>));

describe('Content Details Info', () => {
  it('should render and contain appropriate components', async () => {
    const detail = {
      loading: true,
      name: 'dummy name',
    };

    const detailInfo = {
      dummy_name: 'dummy name',
      dummy_details_field: 'dummy details',
    };

    const displayMap = new Map([
      ['dummy_name', 'Name'],
      ['dummy_details_field', 'Details Field'],
    ]);

    const repositories = [
      {
        id: 155,
        name: 'dummy_name',
        product_id: 1,
        product_name: 'dummy_product',
      },
    ];

    const schema = [
      {
        key: 1,
        tabHeader: 'Details',
        tabContent: (
          <ContentDetailInfo contentDetails={detailInfo} displayMap={displayMap} />
        ),
      },
      {
        key: 2,
        tabHeader: 'Repositories',
        tabContent: (repositories && repositories.length ?
          <ContentDetailRepositories repositories={repositories} /> :
          'No repositories to show'
        ),
      },
    ];

    const wrapper = shallow(<ContentDetails
      contentDetails={detail}
      schema={schema}
    />);

    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(ContentDetailInfo)).toHaveLength(1);
    expect(wrapper.find(ContentDetailRepositories)).toHaveLength(1);
  });
});
