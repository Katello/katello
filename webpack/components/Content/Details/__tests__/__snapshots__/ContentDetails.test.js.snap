// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Details Info should render and contain appropriate components 1`] = `
<div>
  <LoadingState
    loading={true}
    loadingText="Loading"
    timeout={300}
  >
    <ForwardRef
      defaultActiveKey={1}
      id="content-tabs-container"
    >
      <Grid
        bsClass="container-fluid"
        componentClass="div"
        fluid={false}
      >
        <Row
          bsClass="row"
          componentClass="div"
        >
          <Col
            bsClass="col"
            componentClass="div"
            sm={12}
          >
            <Nav
              bsClass="nav nav-tabs"
              id="content-nav-container"
              justified={false}
              pullLeft={false}
              pullRight={false}
              stacked={false}
            >
              <NavItem
                active={false}
                disabled={false}
                eventKey={1}
                key="1"
              >
                <div>
                  Details
                </div>
              </NavItem>
              <NavItem
                active={false}
                disabled={false}
                eventKey={2}
                key="2"
              >
                <div>
                  Repositories
                </div>
              </NavItem>
            </Nav>
          </Col>
        </Row>
        <TabContent
          animation={false}
          bsClass="tab"
          componentClass="div"
          mountOnEnter={false}
          unmountOnExit={false}
        >
          <TabPane
            bsClass="tab-pane"
            eventKey={1}
            key="1"
          >
            <Row
              bsClass="row"
              componentClass="div"
            >
              <Col
                bsClass="col"
                componentClass="div"
                sm={12}
              >
                <ContentDetailInfo
                  contentDetails={
                    Object {
                      "dummy_details_field": "dummy details",
                      "dummy_name": "dummy name",
                    }
                  }
                  displayMap={
                    Array [
                      Array [
                        "dummy_name",
                        "Name",
                      ],
                      Array [
                        "dummy_details_field",
                        "Details Field",
                      ],
                    ]
                  }
                />
              </Col>
            </Row>
          </TabPane>
          <TabPane
            bsClass="tab-pane"
            eventKey={2}
            key="2"
          >
            <Row
              bsClass="row"
              componentClass="div"
            >
              <Col
                bsClass="col"
                componentClass="div"
                sm={12}
              >
                <ContentDetailRepositories
                  repositories={
                    Array [
                      Object {
                        "id": 155,
                        "name": "dummy_name",
                        "product_id": 1,
                        "product_name": "dummy_product",
                      },
                    ]
                  }
                />
              </Col>
            </Row>
          </TabPane>
        </TabContent>
      </Grid>
    </ForwardRef>
  </LoadingState>
</div>
`;
